### ISSUE TITLE
perf: refactor CacheOptimizer/LoopOptimizer + generic-T rank-N FlashAttention + NumOps fast-path — beat PyTorch CPU on small-batch inference

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#294]
## Context

Audit triggered by [`ooples/AiDotNet#1158`](https://github.com/ooples/AiDotNet/issues/1158) (Adopt Tensors' CacheOptimizer / LoopOptimizer in hand-rolled CPU hot paths). The audit refuted the original premise of #1158 and surfaced architectural problems on the Tensors side that block the perf wins those helpers were designed for. This issue consolidates the full set of fixes into one tracked piece of work.

## What's broken today

`CacheOptimizer` and `LoopOptimizer` (in `Engines/Optimization/`) are public API but have **zero internal consumers** — neither this repo nor `ooples/AiDotNet` references them anywhere in `src/`. The current API shape prevents the wins they were designed for:

1. **`CacheOptimizer.TransposeBlocked` / `CopyWithPrefetch` are hardcoded `float[]`** — violates the project's generic-T design philosophy (T everywhere except the GPU boundary, which converts to float/double for device fit).
2. **`LoopOptimizer.Tile2D` / `UnrollBy4` / `UnrollBy8` / `Fuse` / `OptimalOrder2D` / `StripMine` take `Action<int>` delegates** — closure capture forces every call through delegate dispatch, defeats RyuJIT inlining, blocks SIMD auto-vectorization, allocates per call.
3. **Three independent re-implementations of "pick a tile from L1" exist**: `CacheOptimizer.ComputeOptimalTiling`, `LoopOptimizer.DetermineOptimalTileSize`, `MatrixMultiplyHelper.GetBlockSize<T>`. Only the third is generic-T. No consolidation.
4. **No production kernel uses these helpers**: `MatrixMultiplyHelper`, `PackedMatMul`, `FusedDequantMatmulKernels`, `SimdGemmDouble`, `SimdConvHelper`, `CpuFusedOperations`, `FlashAttention2` all roll their own.
5. **`FlashAttention2` is float-only and rank-4-only** — every other shape (rank-2 direct attention, rank-3 single-head batched, rank-5+ video / MoE / Swin windowed) is forced to a parallel implementation. AiDotNet's hand-rolled `FlashAttention<T>` (1014 LOC) exists precisely because Tensors' version doesn't cover those cases.
6. **`NumOps<T>` virtual dispatch dominates element-wise hot loops** — surfaced empirically: a `BinaryCrossEntropyLoss.CalculateDerivative` benchmark showed 4× speedup when the inner loop uses raw `double` arithmetic instead of going through `NumOps`. Every loss function, activation function, and norm kernel pays this tax in the generic path.

Empirical evidence on AMD Ryzen 9 3950X (full benchmarks committed under [AiDotNet `perf/cache-loop-optimizer-1158` branch](https://github.com/ooples/AiDotNet/blob/perf/cache-loop-optimizer-1158/AiDotNetBenchmarkTests/NeuralNetworks/CacheLoopOptimizerBenchmarks.cs)):

| Bench | Naive | Current API | Verdict |
|---|---|---|---|
| Synthetic naive matmul 512×512 | 303 ms | `ComputeOptimalTiling` blocked: **173 ms (43% faster)** | Concept proven, current API can't deliver to real kernels |
| FlashAttention SeqLen=128 | 78 ms | `DetermineOptimalTileSize`: 68 ms (12%, ±14% noise) | Wash |
| FlashAttention SeqLen=512+ | — | — | No diff |
| Embedding scatter Tile2D | 162 μs | **297 μs (84% slower)** | Regression from `Action<int>` closure overhead |
| BCE derivative inner loop | 62 μs | 14 μs (4× via raw double) | NumOps virtual dispatch, not unrolling |

## Goal

Make `CacheOptimizer` + `LoopOptimizer` deliver on their distinct mandates and beat PyTorch CPU on small-batch / inference-style workloads. PyTorch's CPU stack pays 1–5μs per op of Python+ATen dispatch overhead and uses MKL tile sizes baked at MKL build time; .NET virtual call is ~10ns and we can adapt tile sizes at runtime to the actual L1 of the actual chip. That's where the win is — but only with the right API shape.

### Where Tensors *should* beat PyTorch CPU

| Lever | PyTorch reality | Tensors opportunity |
|---|---|---|
| **Op-dispatch overhead** | 1–5μs/op via Python + ATen | ~10ns/op (.NET virtual call). Win: **100×–500×** at small-tensor / batch-1 / inference scenarios |
| **Tile-size adaptivity** | Compiled-in tile sizes from MKL build, tuned for one CPU family at MKL build time | `PlatformDetector.Capabilities.L1CacheSize` at runtime → tile sized for the actual L1 of the actual chip. Meaningful on non-Intel: AMD Zen has 32KB L1d, Apple M-series has 192KB L1d, ARM Neoverse has 64KB |
| **Kernel fusion** | Only via `torch.compile` (TorchInductor + Triton). Default mode launches each op separately | `Fuse` with struct-callback API can be JIT-inlined natively. No torch.compile warmup pass needed |
| **Parallelism** | TBB or OpenMP, GIL-bound on Python side | .NET TPL with no GIL. Multi-threaded matmul/conv on CPU should beat PyTorch CPU outright on >8-core machines |
| **Generic-T** | float/double/half compiled separately | RyuJIT specialization + `NumericFastPath`. Don't maintain 4 copies of every kernel |

### `CacheOptimizer` — memory-layout-focused (the read/write side)

| Method | Purpose |
|---|---|
| `TransposeBlocked<T>` | Cache-line-blocked transpose (memory operation) |
| `CopyWithPrefetch<T>` | `Sse.Prefetch0`-issuing streaming copy (memory latency hiding) |
| `MortonEncode/Decode` | Z-order curves for 2D-locality-preserving 1D layout |
| `ComputeOptimalTiling<T>(m,n,k)` | Matmul-shape tile picker — **single source of truth** for L1-aware block sizing |
| `EstimateCacheMisses` | Pre-flight diagnostic |
| `L1/L2/L3BlockSize` | Cache-hierarchy facts |

Job: "Given a memory layout, do the read/write pattern in a cache-friendly order."

### `LoopOptimizer` — control-flow-focused (the compute side)

| Method | Purpose |
|---|---|
| `Tile2D` / `Tile3D` | Blocked iteration over n-d ranges |
| `UnrollBy4` / `UnrollBy8` | Explicit unrolling so JIT/SIMD can vectorize |
| `OptimalOrder2D(rows, cols, rowMajorAccess)` | Loop interchange based on body's access pattern |
| `StripMine` | Chunked iteration for streaming throughput |
| `Fuse` | Loop fusion: N actions in one pass |
| `ParallelTile2D` | Work-stealing tile scheduler |

Job: "Given a loop's compute body, structure the iteration (tile/unroll/fuse/interchange/parallelize) so the JIT and SIMD have what they need."

The current code muddies the boundary — `LoopOptimizer.DetermineOptimalTileSize` duplicates `CacheOptimizer.ComputeOptimalTiling`, and `CacheOptimizer.TransposeBlocked` muddies generic-T. Resolve.

### CacheOptimizer

```csharp
// Before:
public static void TransposeBlocked(float[] src, float[] dst, int rows, int cols)
public static void CopyWithPrefetch(float[] src, float[] dst, int length)
public static (int, int, int) ComputeOptimalTiling(int m, int n, int k, int elementSize)

// After:
public static void TransposeBlocked<T>(ReadOnlySpan<T> src, Span<T> dst, int rows, int cols)
    where T : unmanaged;
public static void CopyWithPrefetch<T>(ReadOnlySpan<T> src, Span<T> dst)
    where T : unmanaged;
public static (int tileM, int tileN, int tileK) ComputeOptimalTiling<T>(int m, int n, int k)
    where T : unmanaged;  // uses Unsafe.SizeOf<T>()
```

`Sse.Prefetch0` for x86 is in `System.Runtime.Intrinsics.X86`. ARM has `Arm.Prefetch` equivalents. Gate via `Sse.IsSupported` / `ArmBase.IsSupported`.

### LoopOptimizer — struct-callback generics, no delegates

Replace every `Action<int>` / `Action<int, int, int, int>` with **struct-callback generics** so the JIT devirtualizes and inlines. Same pattern `Span<T>.Sort<T,TComparer>` and `Vector<T>` use:

```csharp
public interface ILoopAction { void Invoke(int i); }
public interface ITile2DAction { void Invoke(int i0, int i1, int j0, int j1); }
public interface IInterchangeAction { void Invoke(int i, int j); }

[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static void UnrollBy4<TAction>(int length, TAction action)
    where TAction : struct, ILoopAction
{
    int u = length & ~3;
    for (int i = 0; i < u; i += 4)
    {
        action.Invoke(i);     action.Invoke(i + 1);
        action.Invoke(i + 2); action.Invoke(i + 3);
    }
    for (int i = u; i < length; i++) action.Invoke(i);
}

[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static void Tile2D<TAction>(int rows, int cols, int tileSize, TAction action)
    where TAction : struct, ITile2DAction { /* ... */ }

// Same for Tile3D, OptimalOrder2D, StripMine, Fuse.
```

RyuJIT specializes per concrete struct, devirtualizes `Invoke`, and inlines through. Result: real unrolled loops the SIMD vectorizer can match Intel MKL's inner kernel on.

`DetermineOptimalTileSize<T>` becomes a thin delegate to `CacheOptimizer.ComputeOptimalTiling<T>`. `MatrixMultiplyHelper.GetBlockSize<T>` likewise — single source of truth.

### Generic-T rank-N FlashAttention

Promote `FlashAttention2` → `FlashAttention<T>`. Rename, expand support, NumericFastPath dispatch for the inner softmax math.

```csharp
public static class FlashAttention<T> where T : unmanaged
{
    /// <summary>
    /// Tiled forward. Accepts arbitrary-rank tensors. The LAST TWO dims of
    /// query are [Sq, D]; key is [..., Sk, D]; value is [..., Sk, Dv]. All
    /// preceding dims are batch-like and broadcast across q/k/v under
    /// NumPy-style rules (size-1 dims broadcast against matching values).
    /// Attention-bias's last two dims are [Sq, Sk]; leading dims broadcast
    /// the same way.
    /// </summary>
    public static (Tensor<T> Output, Tensor<T> LogSumExp) Forward(
        Tensor<T> query, Tensor<T> key, Tensor<T> value,
        int blockSizeQ = 64, int blockSizeKV = 64,
        double? scale = null, bool isCausal = false, int queryOffset = 0,
        Tensor<T>? attentionBias = null);

    public static (Tensor<T> GradQ, Tensor<T> GradK, Tensor<T> GradV) Backward(/* ... */);
}
```

Coverage matrix:

| Use case | Rank | Works |
|---|---|---|
| Direct single-batch (debug, research) | 2: `[Sq, D]` | batchProduct=1 |
| Single-head batched | 3: `[B, Sq, D]` | yes |
| Multi-head transformer (canonical FlashAttention-2) | 4: `[B, H, Sq, D]` | yes |
| Video / temporal (TimeSformer, ViViT) | 5: `[B, F, H, Sq, D]` | yes |
| MoE batched experts | 5: `[E, B, H, Sq, D]` | yes |
| Swin windowed | 5+: `[B, W, H, Sq, D]` | yes |
| Arbitrary | N | yes |

**Bias broadcast rule**: last two dims of bias are `[Sq, Sk]`, leading dims broadcast NumPy-style. Subsumes `[Sq,Sk]` ALiBi, `[B,Sq,Sk]` per-batch, `[B,H,Sq,Sk]` per-head, `[1,H,Sq,Sk]` head-only-batch-broadcast, `[B,F,H,Sq,Sk]` video-per-frame — no special cases in the kernel.

**Performance invariant**: rank-4 `[B, H, Sq, D]` runs within 2% of the current rank-fixed `FlashAttention2` (the batchProduct-loop wrapper adds no measurable cost on the canonical shape). SIMD work is unchanged.

**Strided / non-contiguous inputs**: handled at the slice-resolution boundary — either take strides directly if the inner kernel supports it, or materialize a contiguous copy of the `[Sq, D]` slice once before the kernel (cheap relative to attention compute).

**LogSumExp**: shape changes from `float[]` of size `B*H*Sq` to `Tensor<T>` of shape `[...prefix..., Sq]`. Cleaner, generic-T, naturally rank-N.

**NumericFastPath dispatch** (see below): float gets the paper-faithful Dao 2023 kernel; double gets a double-precision kernel for research / scientific computing; other T fall through to a slow generic path with a `[Conditional]`-guarded warning. Same shape as `MathF` vs `Math` in BCL.

### NumOps<T> fast-path for primitives

The 4× win the audit surfaced is real. Pattern:

```csharp
public interface IElementwiseBinaryOp<T> where T : unmanaged
{
    T Apply(T a, T b);
}

public static class NumericFastPath
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static void ElementwiseBinary<T, TOp>(
        ReadOnlySpan<T> a, ReadOnlySpan<T> b, Span<T> dst)
        where T : unmanaged
        where TOp : struct, IElementwiseBinaryOp<T>
    {
        // 1. typeof(T) == typeof(double) fast-path: cast spans, raw double + Vector<double>
        // 2. typeof(T) == typeof(float) fast-path: same with Vector<float>
        // 3. typeof(T) == typeof(Half), int, etc.: SIMD fast-paths where intrinsics exist
        // 4. else: fall through to generic NumOps path with the struct-callback Op
    }

    public static void ElementwiseUnary<T, TOp>(/* ... */);
    public static T ElementwiseReduction<T, TOp>(/* ... */);
}
```

Loss functions, activations, layer-norm reductions all consume this instead of writing the typeof-T branches inline.

## Required wiring (where the helpers actually go to work)

After the API refactor, these production kernels switch to consuming the new APIs:

- `MatrixMultiplyHelper.MultiplyBlocked` → `CacheOptimizer.ComputeOptimalTiling<T>` for block size + `LoopOptimizer.Tile3D` for i-k-j blocked iteration (struct callback)
- `PackedMatMul` pre-pack → `CacheOptimizer.TransposeBlocked<T>` instead of hand-rolled transpose
- `FlashAttention<T>.Forward` / `Backward` → `CacheOptimizer.ComputeOptimalTiling<T>` for `blockSizeQ`/`blockSizeKV` defaults instead of hardcoded 64
- `SimdConvHelper` → `LoopOptimizer.Tile2D` for spatial blocking
- `CpuFusedOperations` → `LoopOptimizer.Fuse` for genuinely-fused element-wise pipelines
- All loss functions, pointwise activations, layer-norm reductions → `NumericFastPath` helper

## Acceptance criteria

1. Zero `Action<int>` / `Action<int,int,int,int>` parameters on any `LoopOptimizer` API.
2. Zero `float[]` parameters on any `CacheOptimizer` API (`Span<T> where T : unmanaged` instead).
3. `CacheOptimizer.ComputeOptimalTiling<T>` is the single L1-aware tile picker — `LoopOptimizer.DetermineOptimalTileSize<T>` and `MatrixMultiplyHelper.GetBlockSize<T>` delegate to it.
4. At least these production kernels actually consume the new APIs: `MatrixMultiplyHelper.MultiplyBlocked`, `PackedMatMul`, `FlashAttention<T>`, `SimdConvHelper`, `CpuFusedOperations`.
5. `NumericFastPath` shared helper exists and is consumed by ≥3 loss functions + ≥3 activation functions + LayerNorm.
6. **Integration benchmarks vs PyTorch CPU on AMD Zen + Intel + ARM** for: matmul (3 sizes), Conv2D (small + big), FlashAttention (3 seq lengths), LayerNorm, BCE derivative. Tensors meets-or-beats PyTorch CPU baseline on ≥80% of cases at batch=1 and batch=32; documented in PR body.
7. No production kernel regresses vs current main on the existing benchmark suite.
8. `FlashAttention<T>` is generic-T (`where T : unmanaged`), accepts arbitrary-rank inputs (rank ≥ 2), where the last two dims are [seq_len, head_dim] for query and last two dims of key/value match the K/V seq layout. All leading dims are broadcast batch-like under NumPy-style rules. Bias broadcasts the same way. AiDotNet can replace its `FlashAttention<T>` call sites — including any rank-5+ video or MoE consumers — with `AiDotNet.Tensors.FlashAttention<T>` and pass the existing test suite plus new rank-2 / rank-5 / broadcast-bias coverage tests.
9. Performance regression test: rank-4 `[B, H, Sq, D]` case runs within 2% of the current rank-fixed `FlashAttention2` implementation — the batchProduct-loop wrapper adds no measurable cost on the canonical shape. Verified via BenchmarkDotNet run as part of the PR.

## Downstream consolidation (separate AiDotNet PR, gated on this Tensors release)

After this Tensors PR ships and a NuGet version is published, a follow-up AiDotNet PR:

- **Deletes** AiDotNet's hand-rolled `FlashAttention<T>` (1014 LOC, parallel implementation). All consumers (CausalTemporalAttention, DiffusionAttention, T5TextConditioner, MultiHeadAttentionLayer, etc.) call `AiDotNet.Tensors.FlashAttention<T>` directly — no wrapper, no `T → float → T` conversion at the boundary.
- AiDotNet's `EmbeddingLayer` and `DeformableConvolutionalLayer` mixed-loop hot paths audited; remaining hand-rolled work pushed to Tensors APIs where feasible.
- AiDotNet loss-function and activation loops switch to `NumericFastPath`.

Result: AiDotNet drops ~1000+ LOC of parallel-implementation code; every AiDotNet attention path automatically inherits future Tensors-side perf work (CUDA kernels, AVX-512 unroll, mixed precision) with zero AiDotNet changes.

## Closes / supersedes

- [`ooples/AiDotNet#1158`](https://github.com/ooples/AiDotNet/issues/1158) (premise refuted by audit; superseded by this issue)
